server.go

```diff
@@ -0,0 +1,207 @@
+
+	return &http.Server{
+		Addr:    httpAddr,
+		Handler: r,
+	}
+}
+
+func readReqBody(r *http.Request) ([]byte, error) {
+	defer r.Body.Close()
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 1048576))
+	if err != nil {
+		log.Errorf("Error occured when reading r.Body: %s", err)
+		return []byte{}, err
+	}
+	return body, nil
+}
+
+type ErrorResponse struct {
+	Error string `json:"error"`
+}
+
+// Custom version of http.Error to support json error messages
+func ErrorRespond(w http.ResponseWriter, errMsg string, code int) {
+	resp := &ErrorResponse{Error: errMsg}
+
+	w.Header().Set(contentType, jsonContentType)
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	w.WriteHeader(code)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Errorf("Error occured when marshalling response: %s", err)
+	}
+}
+
+type CreateEmailAccountResponse struct {
+	Id string `json:"id"`
+}
+
+func CreateEmailAccountHandler(db *sql.DB) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		newAccount := &EmailAccount{}
+		body, err := readReqBody(r)
+		if err != nil {
+			ErrorRespond(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if err := json.Unmarshal(body, newAccount); err != nil {
+			log.Errorf("Error occured when unmarshalling data: %s", err)
+			ErrorRespond(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		err = newAccount.Save(db)
+		if err != nil {
+			ErrorRespond(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		resp := &CreateEmailAccountResponse{
+			Id: newAccount.Id,
+		}
+
+		w.Header().Set(contentType, jsonContentType)
+		w.WriteHeader(http.StatusCreated)
+		if err := json.NewEncoder(w).Encode(resp); err != nil {
+			log.Errorf("Error occured when marshalling response: %s", err)
+			return
+		}
+	}
+}
+
+type SendEmailRequest struct {
+	EmailData  EmailData `json:"email_data"`
+	SecureOnly bool      `json:"secure_only,omitempty"`
+}
+
+func SendEmailHandler(db *sql.DB, emailPool *emailLib.Pool) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := mux.Vars(r)["id"]
+
+		sendEmailReq := &SendEmailRequest{}
+		body, err := readReqBody(r)
+		if err != nil {
+			ErrorRespond(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if err := json.Unmarshal(body, sendEmailReq); err != nil {
+			log.Errorf("Error occured when unmarshalling data: %s", err)
+			ErrorRespond(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// TODO - support returning 500 as well
+		emailAccount, err := GetEmailAccount(db, id)
+		if err != nil {
+			ErrorRespond(w, err.Error(), http.StatusNotFound)
+			return
+		}
+
+		if sendEmailReq.SecureOnly && !emailAccount.HasPubKey {
+			log.Warnf("Failed SecureOnly Email to %s - no pub key", emailAccount.Id)
+			ErrorRespond(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		err = emailAccount.Send(sendEmailReq.EmailData, emailPool)
```

Comment:
@ajvb Send in parallel with goroutines? Or does the email pool sufficiently handle that?

Reply:
I'm not sure, but I think sending in parallel won't really help since the constraint is really on the SMTP server. I think we should move this to a background job (goroutine) where it returns an id for you to check on progress. What do you think of that?